### PR TITLE
ioctl() support for /dev/rshim0/rshim

### PR DESCRIPTION
**Pull request by lsun100**
CUSE doesn't support offset in pread(). This commit adds ioctl()
support to access this file in Linux.

Signed-off-by: Liming Sun <lsun@mellanox.com>

### CODE REVIEW

File: rshim.c [hunk 1]
```diff
@@ -2765,6 +2610,96 @@ static const struct cuse_methods rshim_misc_fops = {
...
+
+/* ioctl message header. */
+typedef struct {
+  uint32_t addr;
+  uint64_t data;
+} __attribute__((packed)) rshim_ioctl_msg;
+
+#ifdef HAVE_RSHIM_FUSE
+enum {
+  RSHIM_IOC_READ = _IO('R', 0),
```
Comment (hselasky) [inline rshim.c]: Size of IOCTL argument must match:
RSHIM_IOC_READ = _IOWR('R', 0, rshim_ioctl_msg)
RSHIM_IOC_WRITE = _IOWR('R', 1, rshim_ioctl_msg)

Comment (lsun100) [reply rshim.c]: Updated. Could you check the new content?

File: rshim.c [hunk 2]
```diff
@@ -2765,6 +2610,139 @@ static const struct cuse_methods rshim_misc_fops = {
...
+  switch (cmd) {
+  case RSHIM_IOC_READ:
+    retval = cuse_copy_in(peer_data, &msg, sizeof(msg));
+    if (retval == CUSE_ERR_NONE) {
+      rc = bd->read_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           &msg.data);
+      if (rc)
+        retval = CUSE_ERR_INVALID
```
Comment (hselasky) [inline rshim.c]: Here you need to do:
else
retval = cuse_copy_out(&msg, peer_data, sizeof(msg));

To get the data back to user-space!

Else this change looks good!

File: rshim.c [hunk 3]
```diff
@@ -2765,6 +2610,139 @@ static const struct cuse_methods rshim_misc_fops = {
...
+typedef struct {
+  uint32_t addr;
+  uint64_t data;
+} __attribute__((packed)) rshim_ioctl_msg;
+
+enum {
+  RSHIM_IOC_READ = _IOWR('R', 0, rshim_ioctl_msg),
+  RSHIM_IOC_WRITE = _IOWR('R', 1, rshim_ioctl_msg),
+};
+
```
Comment (hselasky) [inline rshim.c]: Should probably put this into an own header file: rshim_ioctl.h

File: rshim.c (hunk 2 above)
Comment (lsun100) [reply rshim.c]: Done.

File: rshim.c (hunk 3 above)
Comment (lsun100) [reply rshim.c]: How about leave it to a seperate change later? We probably need to clean up this part a little bit, such as using rshim_fs.c / rshim_fs.h, etc.

File: rshim.c [hunk 4]
```diff
@@ -2765,6 +2610,141 @@ static const struct cuse_methods rshim_misc_fops = {
...
+  case RSHIM_IOC_WRITE:
+    retval = cuse_copy_in(peer_data, &msg, sizeof(msg));
+
+    rc = bd->write_rshim(bd,
+                         (msg.addr >> 16) & 0xF, /* channel # */
+                         msg.addr & 0xFFFF, /* addr */
+                         msg.data);
+    if (rc)
+      retval = CUSE_ERR_INVALID
```
Comment (hselasky) [inline rshim.c]: Missing ";" at end of line.

File: rshim.c [hunk 5]
```diff
@@ -2765,6 +2610,141 @@ static const struct cuse_methods rshim_misc_fops = {
...
+    retval = cuse_copy_in(peer_data, &msg, sizeof(msg));
+    if (retval == CUSE_ERR_NONE) {
+      rc = bd->read_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           &msg.data);
+      if (!rc)
+        retval = cuse_copy_out(&msg, peer_data, sizeof(msg));
+      else
+        retval = CUSE_ERR_INVALID
```
Comment (hselasky) [inline rshim.c]: Missing ";" at end of line.

Comment (lsun100) [reply rshim.c]: Done. I should setup a freebsd building environment :)

File: rshim.c [hunk 6]
```diff
@@ -2765,6 +2610,141 @@ static const struct cuse_methods rshim_misc_fops = {
...
+    break;
+
+  default:
+    fuse_reply_err(req, ENOSYS);
+    break;
+  }
+}
+#endif
+#ifdef HAVE_RSHIM_CUSE
+static void rshim_rshim_ioctl(struct cuse_dev *cdev, int fflags,
```
Comment (hselasky) [inline rshim.c]: static int    and   not     static void.

File: rshim.c [hunk 7]
```diff
@@ -2765,6 +2610,141 @@ static const struct cuse_methods rshim_misc_fops = {
...
+  case RSHIM_IOC_READ:
+    retval = cuse_copy_in(peer_data, &msg, sizeof(msg));
+    if (retval == CUSE_ERR_NONE) {
+      rc = bd->read_rshim(bd,
+                           (msg.addr >> 16) & 0xF, /* channel # */
+                           msg.addr & 0xFFFF, /* addr */
+                           &msg.data);
+      if (!rc)
+        retval = cuse_copy_out(&msg, peer_data, sizeof(msg));
+      else
```
Comment (hselasky) [inline rshim.c]: Same here, missing ";" at end of line below:

File: rshim.c (hunk 3 above)
Comment (hselasky) [reply rshim.c]: No problem.

File: rshim.c (hunk 7 above)
Comment (lsun100) [reply rshim.c]: done

File: rshim.c (hunk 4 above)
Comment (lsun100) [reply rshim.c]: done

**Comment by hselasky** [comment]
Looks good now. You can push!

File: rshim.c (hunk 6 above)
Comment (lsun100) [reply rshim.c]: linux use 'void' prototype in struct cuse_lowlevel_ops{}, which is a little unusual:
void (*ioctl) (fuse_req_t req, int cmd, void *arg, ...)

**Pull request merged**